test/take.cc

```diff
@@ -0,0 +1,117 @@
+#include "stout/take.h"
+
+#include <vector>
+
+#include "gtest/gtest.h"
+#include "stout/collect.h"
+#include "stout/filter.h"
+#include "stout/iterate.h"
+#include "stout/terminal.h"
+
+using stout::eventuals::Collect;
+using stout::eventuals::Filter;
+using stout::eventuals::Iterate;
+using stout::eventuals::TakeFirstN;
+using stout::eventuals::TakeLastN;
+using stout::eventuals::TakeRange;
+
+TEST(Take, IterateTakeLastCollect) {
+  std::vector<int> v = {5, 12, 17, 3};
+
+  auto s = [&]() {
+    return Iterate(v)
+        | TakeLastN(2)
+        | Collect<std::vector<int>>();
+  };
+
+  auto res = *s();
+
+  EXPECT_EQ(17, res.at(0));
+  EXPECT_EQ(3, res.at(1));
+  EXPECT_EQ(2, res.size());
+}
+
+TEST(Take, IterateTakeLastAllCollect) {
+  std::vector<int> v = {5, 12, 17, 3};
+
+  auto s = [&]() {
+    return Iterate(v)
+        | TakeLastN(4)
+        | Collect<std::vector<int>>();
+  };
+
+  auto res = *s();
+
+  EXPECT_EQ(5, res.at(0));
+  EXPECT_EQ(12, res.at(1));
+  EXPECT_EQ(17, res.at(2));
+  EXPECT_EQ(3, res.at(3));
+  EXPECT_EQ(4, res.size());
+}
+
+TEST(Take, IterateTakeRangeCollect) {
+  std::vector<int> v = {5, 12, 17, 20, 22, 1, 1, 1};
+
+  auto s = [&]() {
+    return Iterate(v)
+        | TakeRange(1, 2)
+        | Collect<std::vector<int>>();
+  };
+
+  auto res = *s();
+
+  EXPECT_EQ(2, res.size());
+  EXPECT_EQ(12, res.at(0));
+  EXPECT_EQ(17, res.at(1));
+}
+
+TEST(Take, IterateTakeRangeFilterCollect) {
+  std::vector<int> v = {5, 12, 17, 20};
+
+  auto s = [&]() {
+    return Iterate(v)
+        | TakeRange(1, 2)
+        | Filter([](int x) { return x % 2 == 0; })
+        | Collect<std::vector<int>>();
+  };
+
+  auto res = *s();
+
+  EXPECT_EQ(1, res.size());
+  EXPECT_EQ(12, res.at(0));
+}
+
+TEST(Take, IterateTakeFirstCollect) {
+  std::vector<int> v = {5, 12, 17, 20};
+
+  auto s = [&]() {
+    return Iterate(v)
+        | TakeFirstN(3)
+        | Collect<std::vector<int>>();
+  };
+
+  auto res = *s();
+
+  EXPECT_EQ(3, res.size());
+  EXPECT_EQ(5, res.at(0));
+  EXPECT_EQ(12, res.at(1));
+  EXPECT_EQ(17, res.at(2));
+}
+
+
+TEST(Take, IterateTakeFirstFilterCollect) {
+  std::vector<int> v = {5, 12, 17};
+
+  auto s = [&]() {
+    return Iterate(v)
+        | TakeFirstN(3)
+        | Filter([](int x) { return x % 2 == 1; })
+        | Collect<std::vector<int>>();
+  };
+
+  auto res = *s();
+
+  EXPECT_EQ(2, res.size());
+  EXPECT_EQ(5, res.at(0));
+  EXPECT_EQ(17, res.at(1));
+}
```

Comment:
I'd also like to see tests for what happens if the stream runs out of elements before `N` is reached, for both `TakeRange` and `TakeLastN`.